### PR TITLE
Rename common tests to shared unit tests, and test supported framewor…

### PR DESCRIPTION
**Pull request by drwill-ms**
Rename common tests to shared unit tests, and test supported frameworks (also with devices.client.tests)

### CODE REVIEW

File: iothub/device/tests/ErrorDelegatingHandlerTests.cs [hunk 1]
```diff
@@ -57,7 +56,7 @@ public class ErrorDelegatingHandlerTests
             { typeof(HttpRequestException), () => new HttpRequestException() },
             { typeof(WebException), () => new WebException() },
             { typeof(AmqpException), () => new AmqpException(new Amqp.Framing.Error()) },
-            { typeof(WebSocketException), () => new WebSocketException() },
+            { typeof(WebSocketException), () => new WebSocketException(1) },
```
Comment (drwill-ms) [inline iothub/device/tests/ErrorDelegatingHandlerTests.cs]: no parameter ctor does not exist in all frameworks

File: iothub/device/tests/DeviceClientTests.cs [hunk 2]
```diff
@@ -482,7 +485,7 @@ public async Task DeviceClientSetMethodHandlerSetFirstMethodHandler()
                 actualMethodBody = methodRequest.DataAsJson;
                 actualMethodUserContext = userContext;
                 methodCallbackCalled = true;
-                return Task.FromResult(new MethodResponse(Array.Empty<byte>(), 200));
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs]: Array.Empty<byte>() not supported in all frameworks

File: iothub/device/tests/HsmAuthentication/HttpRequestResponseSerializerTest.cs [hunk 3]
```diff
@@ -219,54 +219,54 @@ public void TestDeserializeResponse_InvalidHeaderSeparator_ShouldThrow()
...
-            Assert.AreEqual(response.Version, Version.Parse("1.1"));
-            Assert.AreEqual(response.StatusCode, System.Net.HttpStatusCode.OK);
-            Assert.AreEqual(response.ReasonPhrase, "OK");
-            Assert.AreEqual(response.Content.Headers.ContentLength, 4);
-            Assert.AreEqual(await response.Content.ReadAsStringAsync(), "Test");
+            Assert.AreEqual(Version.Parse("1.1"), response.Version);
+            Assert.AreEqual(System.Net.HttpStatusCode.OK, response.StatusCode);
+            Assert.AreEqual("OK", response.ReasonPhrase);
+#if !NETCOREAPP1_1
+            Assert.AreEqual(4, response.Content.Headers.ContentLength);
```
Comment (drwill-ms) [inline iothub/device/tests/HsmAuthentication/HttpRequestResponseSerializerTest.cs]: ContentLength is not set in netcoreapp1.1 for some reason.

File: iothub/device/tests/MqttTransportHandlerTests.cs [hunk 4]
```diff
@@ -587,7 +578,7 @@ public async Task MqttTransportHandlerOnErrorCallConnectionClosedListenerOpen()
             Task task = transport.WaitForTransportClosedAsync();
 
             // act
-            transport.OnError(new ApplicationException("Testing"));
```
Comment (drwill-ms) [inline iothub/device/tests/MqttTransportHandlerTests.cs]: Not supported on all frameworks

File: iothub/device/tests/ProductInfoTests.cs [hunk 5]
```diff
@@ -1,74 +1,49 @@
...
         }
 
         [TestMethod]
         public void Extra_DefaultValueIsEmpty()
         {
-            Assert.AreEqual(String.Empty, (new ProductInfo()).Extra);
+            Assert.AreEqual(string.Empty, new ProductInfo().Extra);
         }
 
+#if !NETCOREAPP1_1
```
Comment (drwill-ms) [inline iothub/device/tests/ProductInfoTests.cs]: Not worth getting these tests working on .net core 1.1

File: iothub/device/tests/AmqpTransportHandlerTests.cs [hunk 6]
```diff
@@ -2,23 +2,20 @@
...
-    using Microsoft.Azure.Devices.Client.Transport;
-    using Microsoft.Azure.Devices.Client.Test.ConnectionString;
-    using Microsoft.Azure.Devices.Client.Transport.Amqp;
-
     [TestClass]
     [TestCategory("Unit")]
     public class AmqpTransportHandlerTests
     {
-        const string DumpyConnectionString = "HostName=Do.Not.Exist;SharedAccessKeyName=AllAccessKey;DeviceId=FakeDevice;SharedAccessKey=dGVzdFN0cmluZzE=";
+        private const string DumpyConnectionString = "HostName=Do.Not.Exist;SharedAccessKeyName=AllAccessKey;DeviceId=FakeDevice;SharedAccessKey=dGVzdFN0cmluZzE=";
```
Comment (bikamani) [inline iothub/device/tests/AmqpTransportHandlerTests.cs]: _camelCase

File: iothub/device/tests/CertificateHelper.cs [hunk 7]
```diff
@@ -35,7 +35,9 @@ public static void InstallCertificate(X509Certificate2 certificate, StoreLocatio
             }
             finally
             {
+#if !NETCOREAPP1_1
                 store.Close();
```
Comment (bikamani) [inline iothub/device/tests/CertificateHelper.cs]: context on why is this added?

File: iothub/device/tests/ClientWebSocketTransportTests.cs [hunk 8]
```diff
@@ -1,37 +1,36 @@
...
     [TestCategory("Unit")]
     public class ClientWebSocketTransportTests
     {
-        const string IotHubName = "localhost";
-        const int Port = 12345;
-        static HttpListener listener;
-        static readonly Action<TransportAsyncCallbackArgs> onReadOperationComplete = OnReadOperationComplete;
-        static readonly Action<TransportAsyncCallbackArgs> onWriteOperationComplete = OnWriteOperationComplete;
-        static ClientWebSocketTransport clientWebSocketTransport;
+        private const string IotHubName = "localhost";
```
Comment (bikamani) [inline iothub/device/tests/ClientWebSocketTransportTests.cs]: _camelCase for this and below

File: iothub/device/tests/DeviceClientTests.cs [hunk 9]
```diff
@@ -16,8 +16,8 @@ namespace Microsoft.Azure.Devices.Client.Test
     [TestCategory("Unit")]
     public class DeviceClientTests
     {
-        const string fakeConnectionString = "HostName=acme.azure-devices.net;SharedAccessKeyName=AllAccessKey;DeviceId=dumpy;SharedAccessKey=dGVzdFN0cmluZzE=";
-        const string fakeConnectionStringWithModuleId = "HostName=acme.azure-devices.net;SharedAccessKeyName=AllAccessKey;DeviceId=dumpy;SharedAccessKey=dGVzdFN0cmluZzE=;ModuleId=mod1";
+        private const string fakeConnectionString = "HostName=acme.azure-devices.net;SharedAccessKeyName=AllAccessKey;DeviceId=dumpy;SharedAccessKey=dGVzdFN0cmluZzE=";
```
Comment (bikamani) [inline iothub/device/tests/DeviceClientTests.cs]: _camelCase

File: iothub/device/tests/DeviceClientTests.cs [hunk 10]
```diff
@@ -190,7 +191,7 @@ public async Task DeviceClient_OnMethodCalled_MethodRequestHasEmptyBody()
                 return Task.FromResult(new MethodResponse(Encoding.UTF8.GetBytes("{\"name\":\"ABC\"}"), 200));
             }, "custom data").ConfigureAwait(false);
 
-            var methodRequestInternal = new MethodRequestInternal("TestMethodName", "4B810AFC-CF5B-4AE8-91EB-245F7C7751F9", new MemoryStream(Array.Empty<byte>()), CancellationToken.None);
+            var methodRequestInternal = new MethodRequestInternal("TestMethodName", "4B810AFC-CF5B-4AE8-91EB-245F7C7751F9", new MemoryStream(new byte[0]), CancellationToken.None);
```
Comment (bikamani) [inline iothub/device/tests/DeviceClientTests.cs]: what's the return type here? replace var probably?

File: iothub/device/tests/DeviceClientTests.cs [hunk 11]
```diff
@@ -288,13 +289,15 @@ public async Task DeviceClient_OnMethodCalled_MethodRequestHasEmptyBody_With_Set
             deviceClient.InnerHandler = innerHandler;
 
             bool isMethodHandlerCalled = false;
+#pragma warning disable CS0618 // Type or member is obsolete
```
Comment (bikamani) [inline iothub/device/tests/DeviceClientTests.cs]: Can you explain what is this for?

File: iothub/device/tests/DeviceClientTwinApiTests.cs [hunk 12]
```diff
@@ -11,7 +11,7 @@ namespace Microsoft.Azure.Devices.Client.Test
     [TestCategory("Unit")]
     public class DeviceClientTwinApiTests
     {
-        static string fakeConnectionString = "HostName=acme.azure-devices.net;SharedAccessKeyName=AllAccessKey;DeviceId=dumpy;SharedAccessKey=dGVzdFN0cmluZzE=";
+        private static string fakeConnectionString = "HostName=acme.azure-devices.net;SharedAccessKeyName=AllAccessKey;DeviceId=dumpy;SharedAccessKey=dGVzdFN0cmluZzE=";
```
Comment (bikamani) [inline iothub/device/tests/DeviceClientTwinApiTests.cs]: _camelCase

File: iothub/device/tests/AmqpTransportHandlerTests.cs (hunk 6 above)
Comment (drwill-ms) [reply iothub/device/tests/AmqpTransportHandlerTests.cs]: not for const, and I'm not refactoring the whole file yet.

File: iothub/device/tests/CertificateHelper.cs (hunk 7 above)
Comment (drwill-ms) [reply iothub/device/tests/CertificateHelper.cs]: Close method is not supported in net core app 1.1.

File: iothub/device/tests/ClientWebSocketTransportTests.cs (hunk 8 above)
Comment (drwill-ms) [reply iothub/device/tests/ClientWebSocketTransportTests.cs]: not for const

File: iothub/device/tests/DeviceClientTests.cs (hunk 9 above)
Comment (drwill-ms) [reply iothub/device/tests/DeviceClientTests.cs]: not for const

File: iothub/device/tests/DeviceClientTests.cs (hunk 11 above)
Comment (drwill-ms) [reply iothub/device/tests/DeviceClientTests.cs]: We're calling a method that has an Obsolete attribute. Calling that produces a warning to let the caller know. We are testing the methods still, but I don't want a warning in our build. :)

File: iothub/device/tests/DeviceClientTests.cs (hunk 10 above)
Comment (drwill-ms) [reply iothub/device/tests/DeviceClientTests.cs]: The return type is obvious because we're calling "new MethodRequestInterval". When it is obvious, we prefer "var" as the variable type.

File: iothub/device/tests/DeviceClientTwinApiTests.cs (hunk 12 above)
Comment (drwill-ms) [reply iothub/device/tests/DeviceClientTwinApiTests.cs]: nope

File: iothub/device/tests/ProductInfoTests.cs [hunk 13]
```diff
@@ -124,5 +95,42 @@ public void ToString_ProtocolFormats()
             // HTTP user agent string should not include SQM ID
             Assert.AreEqual(ExpectedHttpUserAgentString(), info.ToString(UserAgentFormats.Http));
         }
+
+        private string ExpectedUserAgentString()
```
Comment (drwill-ms) [inline iothub/device/tests/ProductInfoTests.cs]: Moved below public methods.

File: iothub/device/tests/RetryDelegatingHandlerTests.cs [hunk 14]
```diff
@@ -33,7 +33,7 @@ public async Task RetryDelegatingHandler_OpenAsyncRetries()
             {
                 return ++callCounter == 1
                     ? throw new IotHubException("Test transient exception", isTransient: true)
-                    : Task.CompletedTask;
+                    : TaskHelpers.CompletedTask; ;
```
Comment (abhipsaMisra) [inline iothub/device/tests/RetryDelegatingHandlerTests.cs]: 2 `;`?

File: shared/src/TlsVersions.cs [hunk 15]
```diff
@@ -32,7 +31,7 @@ public class TlsVersions
         public SslProtocols Preferred { get; private set; } = SslProtocols.None;
 
 #if NET451
-        private SecurityProtocolType _net451Protocol = (SecurityProtocolType)PreferredProtocol;
+        private System.Net.SecurityProtocolType _net451Protocol = (System.Net.SecurityProtocolType)PreferredProtocol;
```
Comment (abhipsaMisra) [inline shared/src/TlsVersions.cs]: Why do we want to move `System.Net` out of `using` directive?

File: iothub/device/tests/DeviceClientTests.cs [hunk 16]
```diff
@@ -134,7 +135,7 @@ public async Task DeviceClient_OperationTimeoutInMilliseconds_Equals_0_Open()
             deviceClient.OperationTimeoutInMilliseconds = 0;
 
             var innerHandler = Substitute.For<IDelegatingHandler>();
-            innerHandler.OpenAsync(Arg.Any<CancellationToken>()).Returns(Task.CompletedTask);
```
[the remark this answers is not in the document]
Comment (drwill-ms) [reply iothub/device/tests/DeviceClientTests.cs]: Yes, Task.Completed task wasn't present in all versions, and our multitargetting forces us to use the lowest common denominator.

File: iothub/device/tests/IoTHubClientDiagnosticTest.cs [hunk 17]
```diff
@@ -75,6 +75,8 @@ public void IoTHubClientDiagnostic_DiagId_Test()
             Assert.IsTrue(r.IsMatch(diagId));
         }
 
+#if !NET451
+
```
[the remark this answers is not in the document]
Comment (drwill-ms) [reply iothub/device/tests/IoTHubClientDiagnosticTest.cs]: HttpUtility does not exist in 451

File: iothub/device/tests/RetryDelegatingHandlerTests.cs (hunk 14 above)
Comment (drwill-ms) [reply iothub/device/tests/RetryDelegatingHandlerTests.cs]: Good catch

File: shared/src/TlsVersions.cs (hunk 15 above)
Comment (drwill-ms) [reply shared/src/TlsVersions.cs]: I just hated seeing it gray most of the time, tricking me out as if it were an unused using.
